File: services/gateway-service/src/utils/validation.ts
/**
 * Centralized validation helpers for gateway service.
 */

const SYMBOL_REGEX = /^[A-Z][A-Z0-9\.\-]{0,9}$/;

export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ValidationError";
  }
}

export function normalizeSymbol(raw: string): string {
  const candidate = (raw || "").trim().toUpperCase();
  if (!candidate || !SYMBOL_REGEX.test(candidate)) {
    throw new ValidationError(`Invalid symbol: ${raw}`);
  }
  return candidate;
}

export function parseSymbolsCsv(csv: string): string[] {
  const parts = (csv || "")
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);
  const deduped: string[] = [];
  const seen = new Set<string>();
  for (const part of parts) {
    const normalized = normalizeSymbol(part);
    if (!seen.has(normalized)) {
      seen.add(normalized);
      deduped.push(normalized);
    }
  }
  if (deduped.length === 0) {
    throw new ValidationError("At least one valid symbol is required");
  }
  return deduped;
}

File: shared/python/utils/validation.py
"""
Shared validation helpers for backend services.

Centralizes basic symbol/ticker validation to avoid duplicate regex checks
across FastAPI and streaming services.
"""

from __future__ import annotations

import re
from typing import Iterable, List, Tuple


# Allow indices starting with ^ (e.g. ^GSPC) and increase max length
_SYMBOL_RE = re.compile(r"^[\^A-Z][A-Z0-9\.\-]{0,19}$")


class ValidationError(ValueError):
    """Raised when validation fails for client-provided data."""


def normalize_symbol(symbol: str) -> str:
    """Normalize and validate a single symbol."""
    candidate = (symbol or "").strip().upper()
    if not candidate or not _SYMBOL_RE.match(candidate):
        raise ValidationError(f"Invalid symbol: {symbol}")
    return candidate


def normalize_symbols(symbols: Iterable[str]) -> List[str]:
    """
    Normalize and validate an iterable of symbols, returning a de-duplicated list
    while preserving the original order of first appearance.
    """
    seen = set()
    normalized: List[str] = []
    for raw in symbols:
        cleaned = normalize_symbol(raw)
        if cleaned not in seen:
            seen.add(cleaned)
            normalized.append(cleaned)
    if not normalized:
        raise ValidationError("At least one valid symbol is required")
    return normalized


def parse_symbols_csv(csv_symbols: str) -> List[str]:
    """Parse a comma-separated string of symbols into a validated list."""
    parts = [s for s in (csv_symbols or "").split(",") if s.strip()]
    return normalize_symbols(parts)


